doc/decisions/3_decided/internal_cache.md

```diff
@@ -99,15 +131,19 @@ We make the mmap cache non-optional so that we always have a keyset of configura
 The cache will be used to do change tracking.
 From this keyset, we use `ksBelow` to return the correct keyset.
 
-- Disadvantage: mmap implementation for Windows would be needed
+Disadvantage: mmap implementation for Windows would be needed.
 
-@kodebach wrote:
+The cache should be updated at the end of every kdbGet that was a cache miss.
+So during kdbSet (assuming there is no external modification, i.e. conflict) the on-disk data of the cache should always be up-to-date.
+The idea would be to just read the cached keyset from disk and diff against the current keyset in kdbSet.
 
-> I'm not entirely sure this is possible (@mpranj may know more), but the way I understand it the cache should be updated at the end of every kdbGet that was a cache miss. So during kdbSet (assuming there is no external modification, i.e. conflict) the on-disk data of the cache should always be up-to-date. My idea would be to just read the cached keyset from disk and diff against the current keyset in kdbSet.
->
-> This would mean enabling change tracking also enables the cache (or at least updating the cache, we don't have to use it in kdbGet). If that's not wanted or if the cache data cannot be used directly for some other reason, the same approach could still be used. We'd just have to write the keyset to disk during kdbGet and use it during kdbSet. Since disk space is far less precious than RAM, we could even create separate files for every parent key. If we do go down this route, kdbClose should cleanup the files created by this KDB instance to avoid wasting disk space.
->
-> This approach wouldn't be very performant (since it uses disk IO), but especially if we can use the cache data, it should be pretty easy to do.
+This would mean enabling change tracking also enables the cache (or at least updating the cache, we don't have to use it in kdbGet).
+If that's not wanted or if the cache data cannot be used directly for some other reason, the same approach could still be used.
+We'd just have to write the keyset to disk during kdbGet and use it during kdbSet.
+Since disk space is far less precious than RAM, we could even create separate files for every parent key.
+If we do go down this route, kdbClose should clean up the files created by this KDB instance to avoid wasting disk space.
+
+This approach wouldn't be very performant (since it uses disk IO), but especially if we can use the cache data, it should be pretty easy to do.
```

Comment:
> This is a problem of trying to record the entire discussion in the decision file. 

No, this is a problem of our discussion style. The focus should be on getting a good decision document, not to show off with amazing arguments.

IMHO the main argument is the portability, what about:

```suggestion
Disadvantages:
- It probably will not work on non-POSIX systems, as e.g. [MapViewOfFile](http://msdn.microsoft.com/en-us/library/aa366761%28VS.85%29.aspx) seems to be unreliable and it would be a huge effort to implement that.
- This approach might have a negative effect performance, because additional disk IO might occur.

Advantages:
- If data from the global cache can be reused, it might be easier to implement than the other considered alternatives.
```

Reply:
> > This is a problem of trying to record the entire discussion in the decision file.
> 
> No, this is a problem of our discussion style. The focus should be on getting a good decision document, not to show off with amazing arguments.

If _you_ are trying to show off your amazing arguments, than that _is_ a problem yes. But that's definitely not what I'm doing. I also wasn't trying to blame anybody. I was simply stating a fact. Summarising an organic discussion is not easy. Sometimes you loose context or forget to add extra remarks.

> IMHO the main argument is the portability, what about:

That was not at all my argument. My argument was very simple. The main idea is to store the data **on disk**. This by definition involves disk IO, which _could_ negatively affect performance (if not cached by the OS).

> It probably will not work on non-POSIX systems, as e.g. [MapViewOfFile](http://msdn.microsoft.com/en-us/library/aa366761%28VS.85%29.aspx) seems to be unreliable and it would be a huge effort to implement that.

Please expand on that. How exactly have you come to the conclusion it is unreliable? I would doubt that all memory mapping on Windows is unreliable. Maybe there is a different API, but memory mapped files definitely exist on Windows.

Also I found https://github.com/alitrack/mman-win32 (sadly without a license), so using the Win32 APIs to replace `mmap` seems fairly easy too.